File: src/ctrv_model.cpp
/*
 * Copyright 2022 Leidos
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * Developed by the Human and Vehicle Ensembles (HIVE) Lab at Virginia Commonwealth University (VCU)
 */

#include <cmath>
#include <units.h>
#include "cooperative_perception/ctrv_model.hpp"
#include "cooperative_perception/utils.hpp"
#include "cooperative_perception/units.hpp"

namespace cooperative_perception
{
auto nextState(const CtrvState& state, units::time::second_t time_step) -> CtrvState
{
  using namespace units::literals;

  CtrvState next_state{ state };

  const auto delta_yaw{ state.yaw_rate * time_step };

  if (utils::almostEqual(units::unit_cast<double>(state.yaw_rate), 0.0))
  {
    // Yaw rate of zero (no turning) is a special case. The general case is invalid because it divides by the raw rate.
    // You can't divide by zero.
    next_state.position_x += state.velocity * units::math::cos(state.yaw) * time_step;
    next_state.position_y += state.velocity * units::math::sin(state.yaw) * time_step;
  }
  else
  {
    const auto vel_over_yaw_rate{ state.velocity / state.yaw_rate };
    next_state.position_x +=
        vel_over_yaw_rate * (units::math::sin(state.yaw + delta_yaw) - units::math::sin(state.yaw)) * 1_rad;
    next_state.position_y -=
        vel_over_yaw_rate * (units::math::cos(state.yaw + delta_yaw) - units::math::cos(state.yaw)) * 1_rad;
  }

  next_state.yaw += delta_yaw;

  return next_state;
}

auto nextState(const CtrvState& state, units::time::second_t time_step, const CtrvProcessNoise& noise) -> CtrvState
{
  auto next_state{ nextState(state, time_step) };

  const auto time_step_sq{ units::math::pow<2>(time_step) };
  constexpr auto one_half{ 1.0 / 2.0 };

  next_state.position_x += one_half * noise.linear_acceleration * units::math::cos(state.yaw) * time_step_sq;
  next_state.position_y += one_half * noise.linear_acceleration * units::math::sin(state.yaw) * time_step_sq;
  next_state.velocity += noise.linear_acceleration * time_step;
  next_state.yaw += one_half * noise.angular_acceleration * time_step_sq;
  next_state.yaw_rate += noise.angular_acceleration * time_step;

  return next_state;
}

}  // namespace cooperative_perception

File: scripts/ctrv_model.py
import math

class CtrvState(object):
    def __init__(self, state):
        self.pos_x = state[0]
        self.pos_y = state[1]
        self.velocity = state[2]
        self.yaw = state[3]
        self.yaw_rate = state[4]

    def __str__(self):
        text = f"Current State: \n  x: {self.pos_x}\n  y: {self.pos_y}\n  velocity: {self.velocity}\n  yaw: {self.yaw}\n  yaw_rate: {self.yaw_rate}\n"
        return text

def nextState(state : CtrvState, time_step):
    delta_yaw = state.yaw_rate * time_step;

    if (state.yaw_rate == 0):
        delta_pos_x = state.velocity * math.cos(state.yaw) * time_step;
        delta_pos_y = state.velocity * math.sin(state.yaw) * time_step;

    elif (state.velocity == 0):
        delta_pos_x = 0
        delta_pos_y = 0

    else:
        vel_over_yaw_rate = state.velocity / state.yaw_rate
        delta_pos_x = vel_over_yaw_rate * math.sin(state.yaw + delta_yaw) - math.sin(state.yaw)
        delta_pos_y = -(vel_over_yaw_rate * math.cos(state.yaw + delta_yaw) - math.cos(state.yaw))

    new_state = [state.pos_x + delta_pos_x, state.pos_y + delta_pos_y, state.velocity, state.yaw + delta_yaw, state.yaw_rate]
    return CtrvState(new_state)

# Test for CTRV model:
#  Test CTRV nextState function against pure rotation
# expected: 0, 0, 0, 0.5, 1
state = [0, 0, 0, 0, 1]
time_step = 0.5
n = CtrvState(state)
n = nextState(n, time_step)
print(f"Test 1.) {n}")

# Test for CTRV model:
# Test CTRV nextState function against pure translation
# expected: 0.5, 0, 1, 0, 0
state = [0, 0, 1, 0, 0]
time_step = 0.5
n = CtrvState(state)
n = nextState(n, time_step)
print(f"Test 2.) {n}")

# Test for CTRV model:
# Test CTRV nextState function against mixed rotation and translation
# expected: 0.479425539, 0.122417438, 1, 0.5, 1
state = [0, 0, 1, 0, 1]
time_step = 0.5
n = CtrvState(state)
n = nextState(n, time_step)
print(f"Test 3.) {n}")
